### PR TITLE
Add expired transaction notifications

### PR DESCRIPTION
**Pull request by oxarbitrage**
Closes https://github.com/zcash/zcash/issues/3083

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -528,6 +529,14 @@ static void BlockNotifyCallback(const uint256& hashNewTip)
     boost::thread t(runCommand, strCmd); // thread runs free
 }
 
+static void TrxExpiryNotifyCallback(const std::string& txid)
```
Comment (str4d) [inline src/init.cpp]: Consistent abbreviation:

```suggestion
static void TxExpiryNotifyCallback(const std::string& txid)
```

File: src/ui_interface.h [hunk 2]
```diff
@@ -98,6 +98,9 @@ class CClientUIInterface
 
     /** New block has been accepted */
     boost::signals2::signal<void (const uint256& hash)> NotifyBlockTip;
+
+    /** Transaction expired */
+    boost::signals2::signal<void (const std::string& txid)> NotifyTrxExpiration;
```
Comment (str4d) [inline src/ui_interface.h]:
```suggestion
    boost::signals2::signal<void (const std::string& txid)> NotifyTxExpiration;
```

**Comment by str4d** [comment]
Changes are good. Please squash the fix commit into the main commit, and then I'll merge this 🙂

**Comment by oxarbitrage** [comment]
Force pushed, thank you @str4d

File: src/txmempool.cpp [hunk 3]
```diff
@@ -406,7 +406,7 @@ void CTxMemPool::removeConflicts(const CTransaction &tx, std::list<CTransaction>
     }
 }
 
-void CTxMemPool::removeExpired(unsigned int nBlockHeight)
+std::vector<std::string> CTxMemPool::removeExpired(unsigned int nBlockHeight)
```
Comment (str4d) [inline src/txmempool.cpp]: Hmm.. thinking this over again, we should instead be returning:
```suggestion
std::vector<uint256> CTxMemPool::removeExpired(unsigned int nBlockHeight)
```

File: src/init.cpp [hunk 4]
```diff
@@ -528,6 +529,14 @@ static void BlockNotifyCallback(const uint256& hashNewTip)
     boost::thread t(runCommand, strCmd); // thread runs free
 }
 
+static void TxExpiryNotifyCallback(const std::string& txid)
+{
+    std::string strCmd = GetArg("-txexpirynotify", "");
+
+    boost::replace_all(strCmd, "%s", txid);
```
Comment (str4d) [inline src/init.cpp]: ... and then making this:
```suggestion
    boost::replace_all(strCmd, "%s", txid.GetHex());
```

**Comment by oxarbitrage** [comment]
I think the suggestion is a good one. I added this at https://github.com/zcash/zcash/pull/4343/commits/8f4e438041c633a7246c98f68c463f0ba2a27bb9

We can `git revise` this 2 commits but i wanted to first check. Let me know if i should do that.

**Comment by str4d** [comment]
Yep, that's exactly what I was envisaging :smile:

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Transient failure.

@zkbot retry

**Pull request merged**